FP redundant_pattern (breaks code), "ref" cannot be omitted

````
warning: the `s @ _` pattern can be written as just `s`
    --> src/librustc_resolve/imports.rs:1435:25
     |
1435 |                         ref s @ _ => bug!("unexpected import subclass {:?}", s),
     |                         ^^^^^^^^^ help: try: `s`
     |
     = note: `#[warn(clippy::redundant_pattern)]` on by default
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_pattern
````
using just `s` broke the build:

````
error[E0507]: cannot move out of `directive.subclass` which is behind a shared reference
    --> src/librustc_resolve/imports.rs:1415:37
     |
1415 |                     let msg = match directive.subclass {
     |                                     ^^^^^^^^^^^^^^^^^^ help: consider borrowing here: `&directive.subclass`
...
1435 |                         s => bug!("unexpected import subclass {:?}", s),
     |                         -
     |                         |
     |                         data moved here
     |                         move occurs because `s` has type `imports::ImportDirectiveSubclass<'_>`, which does not implement the `Copy` trait

error: aborting due to previous error

For more information about this error, try `rustc --explain E0507`.
````
it needs to be `ref s`

--- a/clippy_lints/src/misc_early.rs
+++ b/clippy_lints/src/misc_early.rs
@@ -5,8 +5,8 @@ use crate::utils::{
 use if_chain::if_chain;
 use rustc::lint::in_external_macro;
 use rustc_ast::ast::{
-    Block, Expr, ExprKind, GenericParamKind, Generics, Lit, LitFloatType, LitIntType, LitKind, NodeId, Pat, PatKind,
-    StmtKind, UnOp,
+    BindingMode, Block, Expr, ExprKind, GenericParamKind, Generics, Lit, LitFloatType, LitIntType, LitKind, Mutability,
+    NodeId, Pat, PatKind, StmtKind, UnOp,
 };
 use rustc_ast::visit::{walk_expr, FnKind, Visitor};
 use rustc_data_structures::fx::FxHashMap;
@@ -355,7 +355,13 @@ impl EarlyLintPass for MiscEarlyLints {
             }
         }
 
-        if let PatKind::Ident(_, ident, Some(ref right)) = pat.kind {
+        if let PatKind::Ident(left, ident, Some(ref right)) = pat.kind {
+            let left_binding = match left {
+                BindingMode::ByRef(Mutability::Mut) => "ref mut ",
+                BindingMode::ByRef(Mutability::Not) => "ref ",
+                _ => "",
+            };
+
             if let PatKind::Wild = right.kind {
                 span_lint_and_sugg(
                     cx,
@@ -366,7 +372,7 @@ impl EarlyLintPass for MiscEarlyLints {
                         ident.name, ident.name,
                     ),
                     "try",
-                    format!("{}", ident.name),
+                    format!("{}{}", left_binding, ident.name),
                     Applicability::MachineApplicable,
                 );
             }

--- a/tests/ui/patterns.fixed
+++ b/tests/ui/patterns.fixed
@@ -17,4 +17,20 @@ fn main() {
         [x, inside @ .., y] => (), // no error
         [..] => (),
     }
+
+    let mut mutv = vec![1, 2, 3];
+
+    // required "ref" left out in suggestion: #5271
+    match mutv {
+        ref mut x => {
+            x.push(4);
+            println!("vec: {:?}", x);
+        },
+        ref y if y == &vec![0] => (),
+    }
+
+    match mutv {
+        ref x => println!("vec: {:?}", x),
+        ref y if y == &vec![0] => (),
+    }
 }

--- a/tests/ui/patterns.rs
+++ b/tests/ui/patterns.rs
@@ -17,4 +17,20 @@ fn main() {
         [x, inside @ .., y] => (), // no error
         [..] => (),
     }
+
+    let mut mutv = vec![1, 2, 3];
+
+    // required "ref" left out in suggestion: #5271
+    match mutv {
+        ref mut x @ _ => {
+            x.push(4);
+            println!("vec: {:?}", x);
+        },
+        ref y if y == &vec![0] => (),
+    }
+
+    match mutv {
+        ref x @ _ => println!("vec: {:?}", x),
+        ref y if y == &vec![0] => (),
+    }
 }

--- a/tests/ui/patterns.stderr
+++ b/tests/ui/patterns.stderr
@@ -6,5 +6,17 @@ LL |         y @ _ => (),
    |
    = note: `-D clippy::redundant-pattern` implied by `-D warnings`
 
-error: aborting due to previous error
+error: the `x @ _` pattern can be written as just `x`
+  --> $DIR/patterns.rs:25:9
+   |
+LL |         ref mut x @ _ => {
+   |         ^^^^^^^^^^^^^ help: try: `ref mut x`
+
+error: the `x @ _` pattern can be written as just `x`
+  --> $DIR/patterns.rs:33:9
+   |
+LL |         ref x @ _ => println!("vec: {:?}", x),
+   |         ^^^^^^^^^ help: try: `ref x`
+
+error: aborting due to 3 previous errors
 

